proxy/src/proxy.rs

```diff
@@ -966,8 +978,46 @@ impl<S: AsyncRead + AsyncWrite + Unpin> Client<'_, S> {
             params,
             session_id,
             allow_self_signed_compute,
+            endpoint_rate_limiter,
         } = self;
 
+        // Simple rate limiter.
+        //
+        // Check that number of connections to that endpoint is below 300 rps
+        // (purposefully ignoring user name and database names).
+        //
+        // TODO: add a better bucketing here, e.g. not more than 300 requests per second,
+        //       and not more than 1000 requests per 10 seconds, etc. Short bursts of reconnects
+        //       are noramal during redeployments, so we should not block them.
+        if let Some(ep) = creds.get_endpoint() {
+            let endpoint_rate_limiter = endpoint_rate_limiter.clone();
+            let guard = endpoint_rate_limiter
+                .entry(ep)
+                .or_insert_with(|| Arc::new(Mutex::new((chrono::Utc::now().time(), 0))))
+                .clone();
+
+            let check_count = {
+                let mut state = guard.lock();
+                let (date, count) = &mut *state;
+                let now = chrono::Utc::now().time();
+
+                if (now - *date).num_seconds() > 1 {
+                    *date = now;
+                    *count = 0;
+                    *count
+                } else {
+                    *count += 1;
+                    *count
+                }
+            };
+
+            if check_count > 300 {
```

Comment:
This should be configurable. If it's super easy to do, we should do that here. if it's not, it should go in the next release — IIRC we needed to tune the console rate limit a few times after its release.